*: switch to go.mod

otan opened the pull request:
Checklist: https://github.com/cockroachdb/cockroach/issues/50148

----

This PR switches our dependency from `./bin/dep` to `go mod`.

Changes:
* delete Gopkg.toml and Gopkg.lock and replace with go.mod and go.sum.
* Use a patched version of `github.com/goware/modvendor` that can pull
  in imports from packages not in `vendor/modules.txt'.
* delete `c-deps/protobuf/example`, which causes problem with the build.
* change `Makefile` to install vendor directory on `make vendor_update.
* update apd to apd/v2.
* update goavro to goavro/v2.
* fix some lint related tests.
* build: delete references to ./bin/dep. update README instructions.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49447)
<!-- Reviewable:end -->

petermattis reviewed (commented):
Can you pull out the apd update into a separate PR? Most of the file changes are due to that update. Not a big deal if you can't as they seem trivial.

This is very exciting progress! We'll want to make sure this works in all of the environments, such as CI, building release binaries (e.g. `build/builder.sh mkrelease`), building Windows binaries, etc. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-:-M7xgYBH7wOabMMoRe0w:b-qu6l1o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @pbardea, and @petermattis)

___
*[.gitmodules, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7xhaZM0NsLhTC0HHV_:-M7xhaZM0NsLhTC0HHVa:bavhyem) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7985c53f1e6d32abd604f5987b9a44f6b210ee3/.gitmodules#L1)):*
> ```gitconfig
> [submodule "vendor"]
> ```

Removing the vendor submodule seems independent from switching to `go mod`. Is this required for some reason I'm not seeing? Or are the advantages so big that we should do this now rather than waiting?

The part I'm curious about is whether populating `vendor` during CI is significantly slower or more fragile than checking out the submodule.
___
*[Makefile, line 1636 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7xgz7F8Ydy2LZDwzhC:-M7xgz7F8Ydy2LZDwzhD:b-63j3l4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f834515cac1b1035d9fbc06434655a0fe69cb5a8/Makefile#L1636)):*
> ```Makefile
> 	rm -rf artifacts bin $(ARCHIVE) pkg/sql/parser/gen
> 
> vendor: go.sum
> ```

I seem to recall that make is weird when the target is a directory. It might be better for the target to be a file inside of `vendor`. For example, `vendor/modules.txt: go.sum`.
___
*[Makefile, line 1637 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7xfw_lDfICrTrRUTdk:-M7xfw_lDfICrTrRUTdl:b9po2sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f834515cac1b1035d9fbc06434655a0fe69cb5a8/Makefile#L1637)):*
> ```Makefile
> 
> vendor: go.sum
> 	go install github.com/otan-cockroach/modvendor
> ```

Should this be installing into `bin/` like we do for other tools used in the Makefile?

Should you be using `$(GO)` here?


<!-- Sent from Reviewable.io -->

otan commented:
I don't think we can update apd in a separate PR since go mod isn't a thing -- but let me try.





<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea and @petermattis)

___
*[.gitmodules, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7xhaZM0NsLhTC0HHV_:-M7y-MW10RPfma5egwPu:b-wts2pm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7985c53f1e6d32abd604f5987b9a44f6b210ee3/.gitmodules#L1)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Removing the vendor submodule seems independent from switching to `go mod`. Is this required for some reason I'm not seeing? Or are the advantages so big that we should do this now rather than waiting?

The part I'm curious about is whether populating `vendor` during CI is significantly slower or more fragile than checking out the submodule.
</blockquote></details>

Yes. `go mod vendor` deletes the entire directory and then re-adds it from scratch, losing the `.git` references in between. Without some big hacks, there's no point keeping it around.
`go mod` should cache existing versions in `go/pkg/mod`, so hopefully it's a case of copying. it'll be a little slower for sure.
___
*[Makefile, line 1636 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7xgz7F8Ydy2LZDwzhC:-M7y-i9DAY89DYbifJbT:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f834515cac1b1035d9fbc06434655a0fe69cb5a8/Makefile#L1636)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I seem to recall that make is weird when the target is a directory. It might be better for the target to be a file inside of `vendor`. For example, `vendor/modules.txt: go.sum`.
</blockquote></details>

fixed.
___
*[Makefile, line 1637 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7xfw_lDfICrTrRUTdk:-M7y-jfM67gx2BM7bs-9:btyuod1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f834515cac1b1035d9fbc06434655a0fe69cb5a8/Makefile#L1637)):*
> Should this be installing into bin/ like we do for other tools used in the Makefile?




> Should you be using $(GO) here?

fixed.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I have re-read the README.md and I am confused about a few things:

- the doc of `go mod` says that `go get XX@vYYY` will fetch package XX at version YYY *or any later minor version in semver* - how does this achieve version pinning?

- we have *regularly* problems in CI due to github or 3rd party HTTP servers not being available. The solution as presented here suggests that dependencies are fetched 1) upon every fresh TC agent spinning up 2) at arbitrary versions in the past from 3rd party sources, where they may not be available any more.  Does the solution provide a proxy server in our network that preserves these versions? How does that work?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-:-M80_J9v2MEKZJ7mg2A9:b-ardorl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea and @petermattis)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
> the doc of go mod says that go get XX@vYYY will fetch package XX at version YYY or any later minor version in semver - how does this achieve version pinning?

The `go.sum` file contains the pinned version of each module. Unless `go.sum` is modified after the result of a `go get` operation, the build will use the version of each module that's in `go.sum`.

> we have regularly problems in CI due to github or 3rd party HTTP servers not being available. The solution as presented here suggests that dependencies are fetched 1) upon every fresh TC agent spinning up 2) at arbitrary versions in the past from 3rd party sources, where they may not be available any more. Does the solution provide a proxy server in our network that preserves these versions? How does that work?

No, we don't have a proxy server. I believe the devinf team is working on adding one concurrently.

mjibson commented:
Google hosts the canonical Go proxy server at https://proxy.golang.org/ which is used by default since Go 1.13. We could decide to switch to modules and just assume that it'll stay up because we believe they are pretty good at writing and running this kind of infra. As a hedge, it would make sense to maintain a vendor repo in github, but it could be like write-only or something and just there in case the google proxy goes down/away? Kinda not ideal though since if we never use it we wouldn't know if it had problems.

knz commented:
case in point, I tried to fetch dependencies "from scratch" earlier today (for the `registration` server) and it failed, because there's a mix of repository sub-packages that don't exist upstream any more.

It happens to work with proxy.golang.org because they accidentally preserved a copy of those packages "when they were still there", but if proxy.golang.org goes away so does our ability to fetch the historical deps.

I'm not keen on relying on a 3rd party to guarantee we remain able to build historical versions of CockroachDB!

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @petermattis)

___
*[go.mod, line 38 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M8Ju90q9oClljOG8J8y:-M8Ju90r5qA_9keI9o5C:b-syh54x) ([raw file](https://github.com/cockroachdb/cockroach/blob/a534fdd0cbcc3a373ab1bc66bd2189875545657e/go.mod#L38)):*
> ```XML
> 	github.com/cockroachdb/crlfmt v0.0.0-20200116191136-a78e1c207bc0
> 	github.com/cockroachdb/datadriven v0.0.0-20200212141702-aca09668cb24
> 	github.com/cockroachdb/errors v1.2.5-0.20200521081905-32a264b562e5
> ```

some of these need to be updated since this pr was drafted
___
*[pkg/testutils/lint/lint_test.go, line 1661 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M8JuTKi-LNdW9Fl-ZCY:-M8JuTKjDcExtUkGthOy:bo2rph2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a534fdd0cbcc3a373ab1bc66bd2189875545657e/pkg/testutils/lint/lint_test.go#L1661)):*
> ```Go
> 			"SetDetailf",
> 			"SetHintf",
> 			"pgnotice.Newf",
> ```

I'd argue this part of the change belongs to a different, independent PR.
(And should be backported)


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 1 of 1 files at r11, 1 of 1 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @petermattis)

___
*[pkg/testutils/lint/lint_test.go, line 1657 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M8PR7sBEnR8liBfr_e6:-M8PR7sC1RvnXPY2xO3i:bjc2mva) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae9fbf62e72e9ff536cbb069ecb51db7530a919a/pkg/testutils/lint/lint_test.go#L1657)):*
> ```Go
> 			"errors.NewWithDepthf",
> 			"errors.Wrap",
> 			"errors.Wrapf",
> ```

If you're adding arbitrary large chunks of things here, 
please add all the functions from pkg/testutils/lint/passes/fmtsafe/functions.go
specifically those in requireConstFmt

(that's the most complete list we have)


<!-- Sent from Reviewable.io -->

tbg commented:
Re: the historical storage, we can continue updating our `vendored` repo whenever `go.mod` changes via a TC job. Should not be too hard.
Or, and maybe this is preferable in the immediate short term (i.e. as of this PR), we enforce that the vendored directory is embedded (not sub-moduled) here and is up-to-date with go.mod. There was probably a reason we put the vendored files into a separate submodule. I don't remember what it was - it sure is annoying in some ways.

jordanlewis commented:
Tobi, this PR already has dumped the vendored submodule. `go mod` makes it very hard to keep `vendor` submoduled, because it deletes any `.git` directory in there automatically.

jordanlewis commented:
> If you're adding arbitrary large chunks of things here,

Raphael, there's no arbitrariness to the things we're adding. The reason linters have to change is that many of the old analysis packages don't work the same any more given Go modules. I'm not quite sure if you understand how irritating this change is to make - it's very painful and doing it iteratively doesn't feel like a necessary step.

jordanlewis commented:
By the way, these exceptions we've added aren't preventing the linters from failing, so I still don't know what's going on.

`returncheck` is completely broken with go modules. It needs a rewrite to use `packages.Load` instead of `build.Import` which is deprecated and broken in a modules repo. To me this is out of scope and I'd like to disable it completely. Any objections? We have so many linters that it's unclear to me which have the most value.

knz commented:
`returncheck` is important but its current implementation is obsolete and should not be used any more.

Please file an issue to convert `github.com/cockroachdb/returncheck` to a pass in `pkg/testutils/lint/passes` (for use in roachvet)

knz commented:
(and you can disable in the interim)

jordanlewis commented:
Oh, the problem with the `printf` linter is actually that the file you mentioned, @knz, wasn't working *before* this upgrade. The linter was just broken and nobody noticed. CI is failing with many real failures now. I suppose we could just fix them all as part of this PR.

jordanlewis commented:
For example, the following line from `pkg/acceptance/localcluster/tc/tc.go`:

```
		return errors.New(strings.Join(errs, "; "))
```

This is failing the linter because it's a non-constant message, I guess? Seems overly restrictive. Why bother fail a linter for this?

knz commented:
Good catch, let me look into this

tbg commented:
> `returncheck` is completely broken with go modules. It needs a rewrite to use `packages.Load` instead of `build.Import` which is deprecated and broken in a modules repo. To me this is out of scope and I'd like to disable it completely. Any objections? We have so many linters that it's unclear to me which have the most value.

I'm +1 on aggressively disabling any linter that gets in the way of merging this. Just make sure there's an issue for each.

> Tobi, this PR already has dumped the vendored submodule. go mod makes it very hard to keep vendor submoduled, because it deletes any .git directory in there automatically.

👍 I wasn't really considering keeping the submodule, more "are we going to keep vendor checked in" or "do we not check in vendor but instead have a nightly job that snapshots our deps so we can survive if one of them goes away" and in the short term I don't care which one you choose.

knz commented:
>  The linter was just broken and nobody noticed.

Can you tell me how you fixed this?

I am reviewing the linter failures and there's a good mix of legitimate failures (which need code cleanups), missing exclusions in the linter and, possibly, overly restrictive rules.

I can issue a PR with the fixes but I need to reproduce here. Can you suggest what you changed to make it work?

knz commented:
@jordanlewis let's do it the other way: please disable the RoachVet linter in this PR, then I'll work on re-enabling it and fixing the test failures afterwards.

jordanlewis commented:
> Can you tell me how you fixed this?

Unfortunately I have no idea. It happened via the upgrade to Go modules, is all that I know. Perhaps a dependency changed.

> please disable the RoachVet linter in this PR, then I'll work on re-enabling it and fixing the test failures afterwards.

Okay, sounds good. Thank you for your help here.

knz commented:
I think I found it: this PR also upgrades the dependency `golang.org/x/tools` instead of using the version that was pinned in Gopkg.lock.
And that fixes a bunch of linter bugs.

(I know how to proceed further with that, e.g. to fix the linter in at least the v20.1 branch)

Now, there's something here that I don't find comfortable. This PR is now doing two separate things:

- it changes the vendoring system
- and *also* at the same time it changes the version of at least some dependencies

In fact, I don't see that the hashes that were pinned by Gopkg.lock were taken over exactly in Go.mod / .sum. It looks to me instead that the `go mod` command was ran with whatever constraints were present in the Gopkg.toml, resulting in other version choices than the Gopkg.lock.

Would it make sense instead to take over the version pins from Gopkg.lock *exactly* instead?

I would prefer not to combine the work in this PR with the task of checking all the dep upgrades one by one to verify they are not introducing bugs.

jordanlewis commented:
Good point. I'm not sure if this is possible. @otan, did we try to migrate all deps to `go mod` by keeping their shas the same?

tbg commented:
We should spend a bit of time investigating whether we can keep most (I don't think all will be an option) SHAs the same. But we are at the right time in the cycle for an across-the-board dependency bump, so if we find real roadblocks to do so I think we can still proceed. But it will need the usual review that we have done in the past, where we list the new commits we pick up for each dep and sign off on all of them. It should be possible to adapt the old scripts without bending over backwards too much.

tbg commented:
By "a bit of time", I mean parsing today's `Gopkg.lock` and generating from it a set of `go.mod` constraints. If that goes through, great. I'm sure it doesn't and we can see how deep the rabbit hole goes by selectively relaxing incompatible deps' constraints.

otan commented:
I thought go mod in it would preserve it, but it doesn't. When I try force
an old version, it gets deleted. No idea why...

On Thu, 28 May 2020, 7:37 am Tobias Grieger, <notifications@github.com>
wrote:

> We should spend a bit of time investigating whether we can keep most (I
> don't think all will be an option) SHAs the same. But we are at the right
> time in the cycle for an across-the-board dependency bump, so if we find
> real roadblocks to do so I think we can still proceed. But it will need the
> usual review that we have done in the past, where we list the new commits
> we pick up for each dep and sign off on all of them. It should be possible
> to adapt the old scripts without bending over backwards too much.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/49447#issuecomment-635389150>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA32FQZYWOI4QN7HZVCGEN3RTZZJDANCNFSM4NIBDXPA>
> .
>

tbg commented:
Oh, hm, it is claimed that `go mod init` results in "an equivalent build".

https://github.com/golang/go/wiki/Modules#how-to-upgrade-and-downgrade-dependencies

How non-equivalent is it?

tbg commented:
What if you pick one of the deps that now has a different sha, and you run

```
go get chosendep@<oldsha>
```

Is that something we can script and run?

otan commented:
It will get overridden by go mod tidy and go build seems to ignore the old
revision completely.

On Thu, 28 May 2020, 7:43 am Tobias Grieger, <notifications@github.com>
wrote:

> What if you pick one of the deps that now has a different sha, and you run
>
> go get chosendep@<oldsha>
>
> Is that something we can script and run?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/49447#issuecomment-635394781>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA32FQ6ECDFCC5LLFO4AJM3RTZ2BPANCNFSM4NIBDXPA>
> .
>

tbg commented:
I.. don't know enough about vgo to tell you what's going on there, but surely it must be possible to pin a dep to a given sha? Doesn't the `go get asd@sha` change `go.mod` to record that SHA? And you're saying `go mod tidy` just.. throws that out?

knz commented:
Is this suggesting we can't achieve version pinning _at all_? 
(i.e. the moment an upstream dep is upgrade, we'll pick up the update automatically?)

tbg commented:
I mean this must be PEBKAC, vgo def allows you to pin versions and such. @otan want me to give it a go? Can you give me the commands you were running so I can save time?

otan commented:
> Is this suggesting we can't achieve version pinning at all?

I think it means there's a program who's dep contains a different version which is newer than the one we want, so go mod upgrades it implicitly.

> I mean this must be PEBKAC, vgo def allows you to pin versions and such. @otan want me to give it a go? Can you give me the commands you were running so I can save time?

Feel free to checkout the branch and play with it. To do what you were suggesting, I changed the `go.mod`'s require to be the version I wanted, then attempt to `make buildshort` and then afterwards `go mod tidy`. I found that everything gets deleted.

If you were curious about starting from scratch, read `GOMODLOG.md` in the previous attempt: #48853 

We're also chilling in #tmp_go_mod slack channel if you wanted to have a more interactive discussion.

knz commented:
> I think it means there's a program who's dep contains a different version which is newer than the one we want, so go mod upgrades it implicitly.

So I understand that we are currently lacking constraints in `Gopkg.toml`, and so refreshing deps without adding a constraint would update that dep automatically.

This is not great, but it's also not broken - with `vendored` and our dep upgrading policy, we are careful not to bump those deps accidentally.

The current behavior in this PR counts as an "Accident" in this situation.

The way forward I think is this:

1. copy all the current SHAs picked up in Gopkg.lock, and add them as constraints in Gopkg.toml for packages that don't have an exact version constraint already.

2. only then, translate the .toml constraints to the new go mod constraint format.

3. then let the thing pick up the dep.

knz commented:
> The linter was just broken and nobody noticed. CI is failing with many real failures now. 

Fixed in #49660

knz commented:
> Fixed in #49660

that's merged now if you want to rebase

otan commented:
This is ready for review now. There is a link on the PR that describes all go mod implicit upgrades (or downgrades)

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @otan, and @petermattis)

___
*[Makefile, line 313 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9igElb7W7TTehjZ6pr:-M9igElb7W7TTehjZ6ps:bjn100l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/Makefile#L313)):*
> ```Makefile
> vendor_update: bin/.submodules-initialized
> 	rm -rf .vendor.tmp # clean leftovers.
> 	mv vendor .vendor.tmp/
> ```

If something fails in one of the steps below, we'll leave `vendor` in a bad state. Perhaps this should be a `build/update-vendor.sh` script where you can use a `trap 0` to restore the previous state on error.
___
*[Makefile, line 315 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9igxaF8m_OEzFVAsyp:-M9igxaF8m_OEzFVAsyq:b-ihz9hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/Makefile#L315)):*
> ```Makefile
> 	mv vendor .vendor.tmp/
> 	$(GO_INSTALL) -v github.com/goware/modvendor
> 	$(GO) mod vendor
> ```

My understanding is that `go mod vendor` builds the vendor directory, but does not update any of the dependencies (except for ones that were missing). Updating the dependencies is done via `go get -u`. In the Pebble Makefile, I use `go get -u` followed by `go mod vendor` to update the dependencies.
___
*[Makefile, line 316 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9ifqQl1j44Q2pw8rWS:-M9ifqQl1j44Q2pw8rWT:b41bdfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/Makefile#L316)):*
> ```Makefile
> 	$(GO_INSTALL) -v github.com/goware/modvendor
> 	$(GO) mod vendor
> 	./bin/modvendor -copy="**/*.c **/*.h **/*.proto"  -include 'github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis/google/api,github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis/google/rpc,github.com/prometheus/client_model'
> ```

We add `bin` to the front of `PATH` in the Makefile, so I think you can `s,./bin/modvendor,modvendor,g` to match the style for executing built tools used elsewhere in the Makefile.
___
*[build/README.md, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7yA9KVCvIDHizTtWUL:-M7yA9KVCvIDHizTtWUM:b426v41) ([raw file](https://github.com/cockroachdb/cockroach/blob/23aa6da093357df3469bf72542725c5dfda95faf/build/README.md#L89)):*
> ```Markdown
> ## Installing or updating a dependency
> 
> Run `go get -u <dependency>`. To get a specific version, run `go get -u <dependency>@<version>`.
> ```

Can you pin a dependency to a branch, such as what we've done with Pebble on the `release-20.1` branch? I assume yes, but it would be helpful to provide the details here.
___
*[build/README.md, line 93 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9ihM-e00ox473MH3nG:-M9ihM-e00ox473MH3nH:b-aytk3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/build/README.md#L93)):*
> ```Markdown
> Run `go get -u <dependency>`. To get a specific version, run `go get -u <dependency>@<version>`.
> 
> You must then run `make -k vendor_update` to ensure the modules are installed. These changes must
> ```

Perhaps this should be named `vendor_rebuild` or something like that. Are you recommending `-k` here so that we'll do the cleanup if any of the steps encounters an error?NN


<!-- Sent from Reviewable.io -->

knz commented:
The go.mod files in the cockroachdb/errors and sentry-go packages were broken (and invalid). I just fixed them. I hope you can pick up those changes.

otan commented:
> The go.mod files in the cockroachdb/errors and sentry-go packages were broken (and invalid). I just fixed them. I hope you can pick up those changes.

done.

otan reviewed (commented):
Nope - won't build.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-:-M7y2iHH2GvyUmOnPaEU:b-fjoymj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @petermattis)

___
*[go.mod, line 38 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M8Ju90q9oClljOG8J8y:-M9sMlp_-LO_piBdqu2e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a534fdd0cbcc3a373ab1bc66bd2189875545657e/go.mod#L38)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

some of these need to be updated since this pr was drafted
</blockquote></details>

Done.
___
*[Makefile, line 313 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9igElb7W7TTehjZ6pr:-M9sPtM6BYmBYB0e6eYm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/Makefile#L313)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

If something fails in one of the steps below, we'll leave `vendor` in a bad state. Perhaps this should be a `build/update-vendor.sh` script where you can use a `trap 0` to restore the previous state on error.
</blockquote></details>

Done.
___
*[Makefile, line 315 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9igxaF8m_OEzFVAsyp:-M9sPv7-Cp4hq1acVLS8:bon51r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/Makefile#L315)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

My understanding is that `go mod vendor` builds the vendor directory, but does not update any of the dependencies (except for ones that were missing). Updating the dependencies is done via `go get -u`. In the Pebble Makefile, I use `go get -u` followed by `go mod vendor` to update the dependencies.
</blockquote></details>

i've renamed it to `rebuild`.
___
*[Makefile, line 316 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9ifqQl1j44Q2pw8rWS:-M9sQ33vE7XoFFkBemO9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/Makefile#L316)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We add `bin` to the front of `PATH` in the Makefile, so I think you can `s,./bin/modvendor,modvendor,g` to match the style for executing built tools used elsewhere in the Makefile.
</blockquote></details>

Done.
___
*[build/README.md, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M7yA9KVCvIDHizTtWUL:-M9sQWg3EIt3kWcJD61B:b36yjdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/23aa6da093357df3469bf72542725c5dfda95faf/build/README.md#L89)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you pin a dependency to a branch, such as what we've done with Pebble on the `release-20.1` branch? I assume yes, but it would be helpful to provide the details here.
</blockquote></details>

Yep, same invocation. updated.
___
*[build/README.md, line 93 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9ihM-e00ox473MH3nG:-M9sQYOaAnoKTPNKtQ-_:b6btk57) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/build/README.md#L93)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps this should be named `vendor_rebuild` or something like that. Are you recommending `-k` here so that we'll do the cleanup if any of the steps encounters an error?NN
</blockquote></details>

We do `-k` because `make vendor_rebuild` can sometimes build `execgen`, which I don't understand. the `execgen` errors can be ignored.
___
*[pkg/testutils/lint/lint_test.go, line 1661 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M8JuTKi-LNdW9Fl-ZCY:-M9sMefIEr0jRptEz5cc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a534fdd0cbcc3a373ab1bc66bd2189875545657e/pkg/testutils/lint/lint_test.go#L1661)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'd argue this part of the change belongs to a different, independent PR.
(And should be backported)
</blockquote></details>

Done.
___
*[pkg/testutils/lint/lint_test.go, line 1657 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M8PR7sBEnR8liBfr_e6:-M9sMkHK4kAcMuVzYXfL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae9fbf62e72e9ff536cbb069ecb51db7530a919a/pkg/testutils/lint/lint_test.go#L1657)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If you're adding arbitrary large chunks of things here, 
please add all the functions from pkg/testutils/lint/passes/fmtsafe/functions.go
specifically those in requireConstFmt

(that's the most complete list we have)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[build/README.md, line 93 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9ihM-e00ox473MH3nG:-M9ygwf8EDTIRPsHO_Dh:bz7qhgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/build/README.md#L93)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

We do `-k` because `make vendor_rebuild` can sometimes build `execgen`, which I don't understand. the `execgen` errors can be ignored.
</blockquote></details>

Hmm, the `-k` shouldn't be necessary. Should I take a look at what is going on with sometimes building `execgen`?

Nit: `s/vendor_update/vendor_rebuild/g`


<!-- Sent from Reviewable.io -->

otan commented:
___
*[build/README.md, line 93 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9ihM-e00ox473MH3nG:-M9yheX5-LT4t_dBWu5z:b-vij2zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/build/README.md#L93)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Hmm, the `-k` shouldn't be necessary. Should I take a look at what is going on with sometimes building `execgen`?

Nit: `s/vendor_update/vendor_rebuild/g`
</blockquote></details>

Yes please. No idea how that dependency comes out.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @otan)

___
*[build/vendor_rebuild.sh, line 5 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9yhGVm121ILpK_iBWD:-M9yhGVnAjUUV8I0IyD7:bkf0gcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1060fafcfd316bf43bc766b26e5753209f665fe/build/vendor_rebuild.sh#L5)):*
> ```Shell
> set -Eeoux pipefail
> 
> trap restore 1 2 3 6 ERR
> ```

My bash-fu has weakened over time. I think you might be able to accomplish this and more with `trap restore 0`. See https://unix.stackexchange.com/questions/17314/what-is-signal-0-in-a-trap-command#:~:text=If%20a%20sigspec%20is%20EXIT,on%20exit%20from%20the%20shell.&text=As%20%40Mat%20shows%20in%20his,run%20when%20the%20script%20exits.&text=The%20first%20for%20signal%201,second%20because%20the%20script%20exited..
___
*[build/vendor_rebuild.sh, line 14 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9yhg8NC0KEnjsUujlV:-M9yhg8NC0KEnjsUujlW:b1anwkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1060fafcfd316bf43bc766b26e5753209f665fe/build/vendor_rebuild.sh#L14)):*
> ```Shell
> }
> 
> mv vendor .vendor.tmp
> ```

I wonder if it would be worthwhile to use a unique name, either by appending a random number (I know there is a way to do this in bash), or using `mktemp`.


<!-- Sent from Reviewable.io -->

knz commented:
> I wonder if it would be worthwhile to use a unique name, either by appending a random number (I know there is a way to do this in bash)

The word on the street is `$$` (the pid), or `$$$$` in a makefile

otan commented:
sigh, the recent rebase screwed something up and i have no idea.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @petermattis)

___
*[build/vendor_rebuild.sh, line 5 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9yhGVm121ILpK_iBWD:-M9yy7TlCkunYdn6Kt11:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1060fafcfd316bf43bc766b26e5753209f665fe/build/vendor_rebuild.sh#L5)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

My bash-fu has weakened over time. I think you might be able to accomplish this and more with `trap restore 0`. See https://unix.stackexchange.com/questions/17314/what-is-signal-0-in-a-trap-command#:~:text=If%20a%20sigspec%20is%20EXIT,on%20exit%20from%20the%20shell.&text=As%20%40Mat%20shows%20in%20his,run%20when%20the%20script%20exits.&text=The%20first%20for%20signal%201,second%20because%20the%20script%20exited..
</blockquote></details>

Done.
___
*[build/vendor_rebuild.sh, line 14 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9yhg8NC0KEnjsUujlV:-M9yzwCK9g_ImxF-TspO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1060fafcfd316bf43bc766b26e5753209f665fe/build/vendor_rebuild.sh#L14)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I wonder if it would be worthwhile to use a unique name, either by appending a random number (I know there is a way to do this in bash), or using `mktemp`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @otan)

___
*[build/vendor_rebuild.sh, line 5 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9z-JeqBrTv1d9XAgZw:-M9z-JeqBrTv1d9XAgZx:b31uhiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2249d34e18fbf718c14474afd5b7092b5ab5bcc/build/vendor_rebuild.sh#L5)):*
> ```Shell
> set -Eeoux pipefail
> 
> TMP_DIR=.vendor.tmp$(date +"%Y-%m-%d.%H%M%S")
> ```

`TMPDIR` is a special env var in that it controls where temporary files are placed. This variable name is way to close to `TMPDIR` for my comfort. Can we rename this to `tmp_vendor_dir` or something like that. Note that I don't think the current name is problematic, I'm just thinking about the future.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @petermattis)

___
*[build/vendor_rebuild.sh, line 5 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9z-JeqBrTv1d9XAgZw:-M9z-jSmETy5UvAddr2C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2249d34e18fbf718c14474afd5b7092b5ab5bcc/build/vendor_rebuild.sh#L5)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`TMPDIR` is a special env var in that it controls where temporary files are placed. This variable name is way to close to `TMPDIR` for my comfort. Can we rename this to `tmp_vendor_dir` or something like that. Note that I don't think the current name is problematic, I'm just thinking about the future.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
*: switch to go.mod

This PR switches our dependency from `./bin/dep` to `go mod`.

Changes:
* delete Gopkg.toml and Gopkg.lock and replace with go.mod and go.sum.
* Use a patched version of `github.com/goware/modvendor` that can pull
  in imports from packages not in `vendor/modules.txt'.
* delete `c-deps/protobuf/example`, which causes problem with the build.
* change `Makefile` to install vendor directory on `make vendor_update.
* update apd to apd/v2.
* update goavro to goavro/v2.
* fix some lint related tests.
* build: delete references to ./bin/dep. update README instructions.

Release note: None

jordanlewis reviewed (approved):
LGTM! This is a monumental achievement.

:shipit: :shipit: :shipit:

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Very excited to see this go in even if there are a few bumps along the way. :crossed_fingers: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-:-MAaWpex8Hpm02Vu3Ftv:bsh5h44)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

___
*[build/README.md, line 93 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/49447#-M9ihM-e00ox473MH3nG:-MAaWHBl1EjW0waIPasv:b-3bj2yl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d97cbf2527055e7189625327b95e6cd7a18c273a/build/README.md#L93)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yes please. No idea how that dependency comes out.
</blockquote></details>

Well, I didn't get a chance to look at this. Do you mind filing an issue? The use of `make -k` is a bit of a wart, though it might be too difficult to fix with Make.


<!-- Sent from Reviewable.io -->

otan commented:
Filed https://github.com/cockroachdb/cockroach/issues/50609!

otan commented:
bors r=jordanlewis,petermattis

thanks everyone for the help! i'll send an email out.
